lib/internal/perf/usertiming.js

```diff
@@ -0,0 +1,158 @@
+'use strict';
+
+const {
+  ObjectEntries,
+  SafeMap,
+} = primordials;
+
+const {
+  kReadOnlyAttributes,
+  InternalPerformanceEntry,
+  now,
+} = require('internal/perf/perf');
+
+const { enqueue } = require('internal/perf/observe');
+
+const nodeTiming = require('internal/perf/nodetiming');
+
+const {
+  validateNumber,
+  validateObject,
+  validateString,
+} = require('internal/validators');
+
+const {
+  codes: {
+    ERR_INVALID_ARG_VALUE,
+    ERR_INVALID_PERFORMANCE_MARK,
+    ERR_PERFORMANCE_INVALID_TIMESTAMP,
+    ERR_PERFORMANCE_MEASURE_INVALID_OPTIONS,
+  },
+} = require('internal/errors');
+
+const marks = new SafeMap();
+const nodeTimingReadOnlyAttributes =
+  nodeTiming.constructor[kReadOnlyAttributes];
+
+function getMark(name) {
+  if (name === undefined) return;
+  if (typeof name === 'number') {
+    if (name < 0)
+      throw new ERR_PERFORMANCE_INVALID_TIMESTAMP(name);
+    return name;
+  }
+  name = `${name}`;
+  if (nodeTimingReadOnlyAttributes.has(name))
+    return nodeTiming[name];
+  const ts = marks.get(name);
+  if (ts === undefined)
+    throw new ERR_INVALID_PERFORMANCE_MARK(name);
+  return ts;
+}
+
+class PerformanceMark extends InternalPerformanceEntry {
+  constructor(name, options = {}) {
+    name = `${name}`;
+    if (nodeTimingReadOnlyAttributes.has(name))
+      throw new ERR_INVALID_ARG_VALUE('name', name);
+    validateObject(options, 'options');
+    const {
+      detail,
+      startTime = now(),
+    } = options;
+    validateNumber(startTime, 'startTime');
+    if (startTime < 0)
+      throw new ERR_PERFORMANCE_INVALID_TIMESTAMP(startTime);
+    marks.set(name, startTime);
+    super(name, 'mark', startTime, 0, detail);
+    enqueue(this);
+  }
+}
+
+class PerformanceMeasure extends InternalPerformanceEntry {
+  constructor(name, start, duration, detail) {
+    super(name, 'measure', start, duration, detail);
+    enqueue(this);
+  }
+}
+
+function mark(name, options = {}) {
+  return new PerformanceMark(name, options);
+}
+
+function calculateStartDuration(startOrMeasureOptions, endMark) {
+  startOrMeasureOptions ??= 0;
+  let detail;
+  let start;
+  let end;
+  let duration;
+  if (typeof startOrMeasureOptions === 'object' &&
+      ObjectEntries(startOrMeasureOptions).length) {
```

Comment:
Would `Object.keys` be more efficient here (as it creates a 1d array matrix instead of 2d)?
```suggestion
      ObjectKeys(startOrMeasureOptions).length) {
```